File: C++/0690. Employee Importance.cpp
/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution {
public:
    int getImportance(vector<Employee*> employees, int id) {
        map<int, int> emp_import;
        map<int, vector<int>> emp_sub;
        queue<int> q;
        q.push(id);

        for(auto e:employees) {
            emp_import[e->id] = e->importance;
            emp_sub[e->id] = e->subordinates;
        }

        int res = 0;

        while(!q.empty()) {
            int num = q.front();
            q.pop();
            res += emp_import[num];
            for(auto i:emp_sub[num]) {
                q.push(i);
            }
        }
        return res;
    }
};

File: Python3/0690. Employee Importance.py
"""
# Definition for Employee.
class Employee:
    def __init__(self, id: int, importance: int, subordinates: List[int]):
        self.id = id
        self.importance = importance
        self.subordinates = subordinates
"""

class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        emp_dict = {}
        query = [id]

        for employee in employees:
            emp_dict[employee.id] = [employee.importance, employee.subordinates]

        res = 0
        while query:
            num = query.pop(0)
            res += emp_dict[num][0]
            query += emp_dict[num][1]
        
        return res
